File: codingame/python/StockExLoss.py
import unittest

def max_loss(stk_prices: list) -> int:
    ''' actuallt it is maximum drawdown'''
    peak = floor = res = 0
    prev = -1
    for v in stk_prices:
        if prev == -1 or (prev < v and peak < v):
            peak = v
            floor = 0
        elif prev > v and (floor == 0 or floor > v):
            floor = v
            res = max(res, peak - floor)
        prev = v
    return -res

def calcu_bp(numerator, denominator) -> int:
    # if delta * 10000 // denominator, it rounds down to int
    return round(numerator * 10000 / denominator)

# max drawdown(MDD) in basis point(0.01%)
def max_drawdown_bp(stk_prices: list) -> int:
    peak = 0.0  # global high
    trough = 0.0 # subsequent low after peak
    res = 0
    prev = -1
    for v in stk_prices:
        if prev == -1 or (prev < v and peak < v):
            peak = v
            trough = 0
        elif prev > v and (trough == 0 or trough > v):
            trough = v
            res = max(res, calcu_bp(peak - trough, peak))
        prev = v
    return -res

def max_drawdown_duration(stk_prices: list) -> int:
    peak = 0.0  # global high
    peak_idx = -1
    res = 0 # in days
    prev = -1.0
    in_dd = False # in drawdown
    for i in range(len(stk_prices)):
        v = stk_prices[i]
        if prev == -1 or (prev <= v and peak <= v):
            # in order to measure the duration, only narrow brim is accepted, so it is '<='
            if (in_dd):
                res = max(res, i - peak_idx)
            peak = v
            peak_idx = i
            in_dd = False
        elif prev > v or (in_dd and peak > v):
            res = max(res, i - peak_idx)
            in_dd = True
        prev = v
    return res

class TestStockExLoss(unittest.TestCase):
    # each test case must start with 'test_'
    def test_simple_int(self):
        prices = [3, 2, 4, 2, 1, 5]
        self.assertEqual(-3, max_loss(prices))
    
    def test_mdd_bp(self):
        prices = [10.12, 10.23, 10.91, 10.11, 9.61, 10.20, 10.98, 11.35, 11.81, 11.53, 10.91, 10.51, 10.76]
        self.assertEqual(-1192, max_drawdown_bp(prices)) # 10.91 -> 9.61

    def test_mdd_bp_no_down(self):
        prices = [9.0, 9.3, 9.6, 9.8, 10.2]
        self.assertEqual(0, max_drawdown_bp(prices))

    def test_mdd_dura_multi(self):
        prices = [10.1, 9.0, 9.3, 9.8, 10.2, 9.7, 9.6, 9.9]
        self.assertEqual(4, max_drawdown_duration(prices)) # 10.1 -> 10.2

    def test_mdd_dura_long(self):
        prices = [10.12, 10.23, 10.91, 10.11, 9.61, 10.20, 10.54, 10.81, 11.15, 11.41, 11.33, 10.91, 10.51, 10.76]
        self.assertEqual(6, max_drawdown_duration(prices)) # 10.91 -> 11.15

    def test_mdd_dura_flat_days(self):
        prices = [10.1, 10.1, 9.0, 9.3, 9.8, 10.1, 10.1, 10.2, 9.7, 9.6, 9.9]
        self.assertEqual(4, max_drawdown_duration(prices)) # 10.1[1] -> 10.1[5]

if __name__ == '__main__':
    unittest.main()

File: codingame/cpp/src/StockExLoss.cpp
/*
* Input
N stock values arranged in order of trading days.
The values are integers.

Output
The maximal loss p, expressed negatively if there is a loss, otherwise 0.
*/

#include <iostream>
#include <algorithm>
#include <cmath> // fabs()
#include <cfloat>
using namespace std;

class Solution
{
    public:
    // return absolute value of maximal loss
    int max_loss_abs(int* stock_prices, int n)
    {
        int peak = 0; // global high
        int trough = 0;
        int res = 0;
        int prev = -1;
        for (int i = 0; i < n; i++) {
            int v = stock_prices[i];
            if (prev == -1 || (prev < v && peak < v))
            {
                peak = v;
                trough = 0;
            } else if (prev > v && (trough == 0 || trough > v))
            {
                trough = v;
                res = max(res, peak - trough);
            }
            prev = v;
        }
        return 0 - res;
    }
    // maximal loss in base point(BP), or 0 if no loss; 1 BP is 0.01%
    // actually, it is equivalent to maximum drawdown(MDD)
    int max_drawdown_bp(float* stock_prices, int n)
    {
        float peak = 0.0f; // global high
        float trough = 0.0f; // subsequent low after peak
        int res = 0; // unit in BP which is 0.01%
        float prev = -1;
        for (int i = 0; i < n; i++) {
            float v = stock_prices[i];
            if (prev == -1 || (prev < v && peak < v))
            {
                peak = v;
                trough = 0;
            } else if (prev > v && (trough == 0 || trough > v))
            {
                trough = v;
                res = max(res, calcu_bp(peak, trough));
            }
            prev = v;
        }
        return 0 - res;

    }
    // maximal drawdown duration in days, or 0 if no loss
    int max_drawdown_duration(float* stock_prices, int n)
    {
        float peak = 0.0f; // global high
        float prev = -1.0f;
        int peak_idx = -1;
        int res = 0; // in days
        bool in_drawdown = false;
        for (int i = 0; i < n; i++) {
            float v = stock_prices[i];
            if (prev == -1
                || (is_float_lt_equal(prev, v) && is_float_lt_equal(peak, v)))
            {
                // in order to measure the duration, only narrow brim is accepted, so it is '<='
                if (in_drawdown)
                {
                    res = max(res, i - peak_idx); // end of drawdown duration!
                }
                // update for up trend
                peak = v;
                peak_idx = i;
                in_drawdown = false;
            } else if (prev > v || (in_drawdown && peak > v))
            {
                res = max(res, i - peak_idx);
                in_drawdown = true;
            }
            prev = v;
        }
        return res;
    }

    // accumulation of all the available profit in absolute value
    float total_profit_abs(float* stock_prices, int n)
    {
        float res = 0.0f;
        for (int i = 1; i < n; i++) {
            float diff = round_to_n_decimals(stock_prices[i] - stock_prices[i - 1], 2); // 0.01 precision
            if (diff > 0.0) res += diff;
        }
        return res;
    }

    private:
    // calculate loss in BP
    inline int calcu_bp(float peak, float trough)
    {
        // implicit convert float to int rounds down to 0
        // std::round() round to nearest integer
        return round((peak - trough) * 10000 / peak);
    }

    // round to given decimal precision
    float round_to_n_decimals(float value, int n) {
        double multiplier = std::pow(10.0, n);
        return std::round(value * multiplier) / multiplier;
    }

    // compare whether two floats are equal within an epsilon
    inline bool are_floats_equal(float a, float b, float epsilon = FLT_EPSILON) {
        return fabs(a - b) < epsilon;
    }

    // if a <= b
    bool is_float_lt_equal(float a, float b, float epsilon = FLT_EPSILON) {
        return ((b - a) > epsilon) || (fabs(a - b) < epsilon);
    }
};